braindecode/datasets/base.py

```diff
@@ -150,32 +150,47 @@ class WindowsDataset(BaseDataset):
 
     Parameters
     ----------
-    windows : mne.Epochs
+    windows : mne.Raw or mne.Epochs (Epochs is outdated)
         Windows obtained through the application of a windower to a BaseDataset
         (see `braindecode.datautil.windowers`).
     description : dict | pandas.Series | None
         Holds additional info about the windows.
     transform : callable | None
         On-the-fly transform applied to a window before it is returned.
     targets_from : str
-        Defines whether targets will be extracted from mne.Epochs metadata or mne.Epochs `misc`
+        Defines whether targets will be extracted from  metadata or from `misc`
         channels (time series targets). It can be `metadata` (default) or `channels`.
+    last_target_only : bool
+        If targets are obtained from misc channels whether all targets if the entire
+    (compute) window will be returned or only the last target in the window.
+    metadata : pandas.DataFrame
+        Dataframe with crop indices, so `i_window_in_trial`, `i_start_in_trial`, `i_stop_in_trial`
+        as well as `targets`.
     """
+
     def __init__(self, windows, description=None, transform=None, targets_from='metadata',
-                 last_target_only=True):
+                 last_target_only=True, metadata=None, ):
         self.windows = windows
+        if hasattr(self.windows, 'metadata'):
+            metadata = self.windows.metadata
+        else:
+            if metadata is None:
+                raise ValueError("Need to supply metadata if it is not given as part of mne epochs")
+            self.metadata = metadata
+
         self._description = _create_description(description)
+
         self.transform = transform
         self.last_target_only = last_target_only
         if targets_from not in ('metadata', 'channels'):
             raise ValueError('Wrong value for parameter `targets_from`.')
         self.targets_from = targets_from
 
-        self.crop_inds = self.windows.metadata.loc[
-            :, ['i_window_in_trial', 'i_start_in_trial',
-                'i_stop_in_trial']].to_numpy()
+        self.crop_inds = metadata.loc[
```

Comment:
how about keeping it in the form that is needed from the beginning -> I don't know how long will be `.to_list()` call but if it's done at every `__getitem__` this could have some influence

Reply:
This is actually what is done here as this is not done in `__getitem__` no?

Reply:
I meant we're calling `.tolist()` in `__getitem__` in this line https://github.com/braindecode/braindecode/pull/515/files#diff-0e3c9da7298e0ba7355437c41fcfd2d4ef72d0dddb5eb06e857a25499c5d3e64R215
maybe it can be done in the `__init__` and then just accessed

Reply:
It is a list of three items so probably okay I would say but maybe this can be benchmarked.

Reply:
yeah should be fine